src/adagrams.js

```diff
@@ -1,15 +1,100 @@
 export const drawLetters = () => {
-  // Implement this method for wave 1
-};
+  let alphabetPool = 'aaaaaaaaaabbccddddeeeeeeeeeeeeffggghhiiiiiiiiijkllllmmnnnnnnooooooooppqrrrrrrssssttttttuuuuvvwwxyyz'
+  alphabetPool = alphabetPool.toUpperCase()
+  let i = 0;
+  let playerHand = []
+  
+  do {
+    i++;
+    let singleLetter = alphabetPool.charAt(Math.random() * alphabetPool.length);
+    playerHand.push(singleLetter);
+    alphabetPool = alphabetPool.replace(singleLetter, '');
+  } while (i < 10);
+
+  return playerHand;
+}
 
 export const usesAvailableLetters = (input, lettersInHand) => {
-  // Implement this method for wave 2
+  let playerHand = {}
+  for (let i of lettersInHand) {
+    if (i in playerHand) {
+      playerHand[i] += 1
+    } else {
+      playerHand[i] = 1
+    }
+  } 
+
+  for (let j of input) {
+    if (!(j in playerHand)) {
+      return false;
+    } else {
+      playerHand[j] -= 1
+    }
+  }
+
+  return Object.values(playerHand).every(
+    value => value >= 0
+  )
 };
 
 export const scoreWord = (word) => {
-  // Implement this method for wave 3
+  const scoreChart = {
+    1: ['A', 'E', 'I', 'O', 'U', 'L', 'N', 'R', 'S', 'T'],
+    2: ['D', 'G'],
+    3: ['B', 'C', 'M', 'P'],
+    4: ['F', 'H', 'V', 'W', 'Y'],
+    5: ['K'],
+    8: ['J', 'X'],
+    10: ['Q', 'Z']
+  };
+  
+  let wordScore = 0;
+  for (let i = 0; i < word.length; i++) {
+    for (const [key, value] of Object.entries(scoreChart)) {
+      for (const v of value) {
+        if (word[i].toUpperCase() === v) {
+          wordScore += Math.floor(key)
+        }
+      }
+    }
+  };
+  if (word.length >= 7) {
+    wordScore += Math.floor(8) 
+  };
+  
+  return wordScore;
 };
 
+
+
 export const highestScoreFrom = (words) => {
-  // Implement this method for wave 4
+  
+  let wordScores = {};
+ 
+
+  for (let word of words) {
+    wordScores[word] = scoreWord(word);
+  }
+
+
+  const maxScore = Math.max(...Object.values(wordScores));
+
+
+  let allMaxScoringWords = Object.keys(wordScores).filter(
+    (key) => wordScores[key] === maxScore
+  );
```

Comment:
Great start! This gets us to the correct answer, but requires going through the list of words multiple times. There is a way to just go through the list once. This requires having an object that keeps track of the current highest-scoring word and checking it against the next word to see if it should be replaced until you make it through the whole list.

This works but you could simplify this by creating one variable to hold the high scoring data. Using two variables takes up more memory and most primitives (booleans, numbers, null, undefined, etc.) are constant (O(1)) space. This is because, no matter if true or false, or 1 or 1,000, the amount of space consumed stays constant. Reference types (arrays and objects) — Reference types require O(n) space as well. ’n’ being the length for an array, or in the case of an object, ’n’ being the amount of keys. Reference types are O(n) because the more elements in an array, or the more keys in an object, the more space they consume.

Consider checking if the current word is shorter than the current highest scoring word 👍🏾

Consider also creating a data structure to hold both the winning word and score together and setting that value to falsey and then reassigning based on some condition. Sort of like the get_highest_rated PSE 😄